src/algebra/free_algebra.lean

```diff
@@ -0,0 +1,293 @@
+-- commutative additive semigroup
+| add_assoc {a b c : pre R X} : rel (a + b + c) (a + (b + c))
+| add_comm {a b : pre R X} : rel (a + b) (b + a)
+| zero_add {a : pre R X} : rel (0 + a) a
+-- multiplicative monoid
+| mul_assoc {a b c : pre R X} : rel (a * b * c) (a * (b * c))
+| one_mul {a : pre R X} : rel (1 * a) a
+| mul_one {a : pre R X} : rel (a * 1) a
+-- distributivity
+| left_distrib {a b c : pre R X} : rel (a * (b + c)) (a * b + a * c)
+| right_distrib {a b c : pre R X} : rel ((a + b) * c) (a * c + b * c)
+-- other relations needed for semiring
+| zero_mul {a : pre R X} : rel (0 * a) 0
+| mul_zero {a : pre R X} : rel (a * 0) 0
+-- compatibility
+| add_compat_left {a b c : pre R X} : rel a b → rel (a + c) (b + c)
+| add_compat_right {a b c : pre R X} : rel a b → rel (c + a) (c + b)
+| mul_compat_left {a b c : pre R X} : rel a b → rel (a * c) (b * c)
+| mul_compat_right {a b c : pre R X} : rel a b → rel (c * a) (c * b)
+
+end free_algebra
+
+/--
+The free algebra for the type `M` over the commutative semiring `R`.
+-/
+def free_algebra := quot (free_algebra.rel R X)
+
+namespace free_algebra
+
+local attribute [instance]
+  pre.has_coe_generator pre.has_coe_semiring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar
+
+instance : semiring (free_algebra R X) :=
+{ add := quot.map₂ (+) (λ _ _ _, rel.add_compat_right) (λ _ _ _, rel.add_compat_left),
+  add_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.add_assoc },
+  zero := quot.mk _ 0,
+  zero_add := by { rintro ⟨⟩, exact quot.sound rel.zero_add },
+  add_zero := begin
+    rintros ⟨⟩,
+    change quot.mk _ _ = _,
+    rw [quot.sound rel.add_comm, quot.sound rel.zero_add],
+  end,
+  add_comm := by { rintros ⟨⟩ ⟨⟩, exact quot.sound rel.add_comm },
+  mul := quot.map₂ (*) (λ _ _ _, rel.mul_compat_right) (λ _ _ _, rel.mul_compat_left),
+  mul_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.mul_assoc },
+  one := quot.mk _ 1,
+  one_mul := by { rintros ⟨⟩, exact quot.sound rel.one_mul },
+  mul_one := by { rintros ⟨⟩, exact quot.sound rel.mul_one },
+  left_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.left_distrib },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.right_distrib },
+  zero_mul := by { rintros ⟨⟩, exact quot.sound rel.zero_mul },
+  mul_zero := by { rintros ⟨⟩, exact quot.sound rel.mul_zero } }
+
+instance : inhabited (free_algebra R X) := ⟨0⟩
+
+instance : has_scalar R (free_algebra R X) :=
+{ smul := λ r a, quot.lift_on a (λ x, quot.mk _ $ ↑r * x) $
+  λ a b h, quot.sound (rel.mul_compat_right h) }
+
+instance : algebra R (free_algebra R X) :=
+{ to_fun := λ r, quot.mk _ r,
+  map_one' := rfl,
+  map_mul' := λ _ _, quot.sound rel.mul_scalar,
+  map_zero' := rfl,
+  map_add' := λ _ _, quot.sound rel.add_scalar,
+  commutes' := λ _, by { rintros ⟨⟩, exact quot.sound rel.central_scalar },
+  smul_def' := λ _ _, rfl }
+
+/--
+The canonical function `X → free_algebra R X`.
+-/
+def ι : X → free_algebra R X :=
+λ m, quot.mk _ m
```

Comment:
```suggestion
def ι : X → free_algebra R X := λ m, quot.mk _ m
```

Reply:
I still don't really know what the preferred spacing should be in this case, but I personally prefer a more compact style.